### PR TITLE
Make it possible to build with c++20 and switch to new Key4hep build action

### PR DESCRIPTION
**Pull request by tmadlener**
BEGINRELEASENOTES
- Allow to build with c++20
- Switch to new key4hep build action and remove old workflow

ENDRELEASENOTES

### CODE REVIEW

File: .github/workflows/key4hep-build.yaml [hunk 1]
```diff
@@ -3,6 +3,7 @@ name: Key4hep build
 on:
   push:
   workflow_dispatch:
+  pull_request:
```
Comment (jmcarcell) [inline .github/workflows/key4hep-build.yaml:6]: This will get overwritten. Why is this needed? With the `push:` they run whenever there is a push in a branch, in a PR or not, `pull_request:` will duplicate the jobs for a PR since they will run in the PR and after push meaning it takes longer to find workers and for all the jobs to run

**Comment by jmcarcell** [comment]
The spack recipe was missing having the standard specified so it was being built with C++17 instead of C++20 for the nightlies

File: .github/workflows/key4hep-build.yaml (hunk 1 above)
Comment (tmadlener) [inline .github/workflows/key4hep-build.yaml:6]:
```suggestion
```

Comment (tmadlener) [reply .github/workflows/key4hep-build.yaml]: :+1: removed

Comment (tmadlener) [reply .github/workflows/key4hep-build.yaml]: Without the `pull_request` they will not report back to the PR. The `push` hook only runs them on the branch (which is in my fork). Maybe `push: master` and `pull_request:` is the way to go?

Comment (jmcarcell) [reply .github/workflows/key4hep-build.yaml:6]: Yes I think that's it. The repos have to have the same name for the default branch though but in this case it's true for key4hep

**Review by jmcarcell — APPROVED** [review]

File: .github/workflows/key4hep-build.yaml (hunk 1 above)
Comment (jmcarcell) [reply .github/workflows/key4hep-build.yaml:6]: Ah no I just remembered why I didn't go for that. With that, the workflows don't run in your workflow when you push to a different branch other than the default one and some people don't like that. I couldn't find a way of having it
- Run once in PRs
- Run also when pushing to a fork
- Report in the PR page
The closest is the first two without the third one with only `push:`

Comment (tmadlener) [reply .github/workflows/key4hep-build.yaml:6]: OK. But the one that we definitely need is the reporting, right? Otherwise we cannot use auto-merge with required checks. 

I also didn't find a way to restrict the `push` to only the main repository, which I guess would be nice, because then there would be no duplicate reporting of failures.
